lib/plugins/aws/package/lib/mergeIamTemplates.js

```diff
@@ -38,19 +37,30 @@ module.exports = {
         );
       });
 
+    const { iam } = this.serverless.service.provider;
+    const iamRole = iam ? iam.role : {};
+
     // resolve early if provider level role is provided
+    if (typeof iamRole === 'string') {
+      return;
+    }
+
     if ('role' in this.serverless.service.provider) {
-      return BbPromise.resolve();
+      this.serverless._logDeprecation(
+        'PROVIDER_IAM_SETTINGS',
+        'Starting with version 3.0.0, "provider.role" property will be replaced with "provider.iam.role"'
+      );
+      return;
     }
 
     // resolve early if all functions contain a custom role
-    const customRolesProvided = [];
-    this.serverless.service.getAllFunctions().forEach((functionName) => {
+    const customRolesProvided = this.serverless.service.getAllFunctions().every((functionName) => {
```

Comment:
Let's not mix plain refactors into this PR, let's propose them with other PR (we can merge them right away)

This PR touches sensitive part and best if we only propose the code that is specific to proposed change